server/services/prisonerInformation.js

```diff
@@ -0,0 +1,84 @@
+const { logger } = require('../utils/logger');
+
+function createPrisonMapFrom(prisonDetailsResponse) {
+  return prisonDetailsResponse
+    .filter(p => p !== null)
+    .reduce((a, p) => {
+      a.set(p.agencyId, p.longDescription);
+      return a;
+    }, new Map());
+}
+
+function formatTransactionsWith(transactions, prisonDetails) {
+  return transactions.map(t => ({
+    ...t,
+    prison: prisonDetails.has(t.agencyId)
+      ? prisonDetails.get(t.agencyId)
+      : t.agencyId,
+  }));
+}
+
+class PrisonerInformationService {
+  constructor({ prisonApiRepository }) {
+    this.prisonApi = prisonApiRepository;
+  }
+
+  async getTransactionInformationFor(user, accountCode, fromDate, toDate) {
+    if (!user || !accountCode || !fromDate || !toDate) {
+      throw new Error('Incorrect parameters passed');
+    }
+
+    try {
+      const [transactionsResponse, balancesResponse] = await Promise.all([
+        this.prisonApi.getTransactionsFor(
+          user.prisonerId,
+          accountCode,
+          fromDate,
+          toDate,
+        ),
+        this.prisonApi.getBalancesFor(user.bookingId),
+      ]);
+
+      const balances = balancesResponse || {
+        error: 'We are not able to show your balances at this time',
+      };
+
+      if (transactionsResponse) {
+        const listOfPrisons = Array.from(
+          new Set(transactionsResponse.map(t => t.agencyId)),
+        );
+        const prisonDetailsResponse = await Promise.all(
+          listOfPrisons.map(p => this.prisonApi.getPrisonDetailsFor(p)),
```

Comment:
rather than call get prison details for each prison, you could retrieve all prisons in one go.

there's a couple of endpoints: 
https://api.prison.service.justice.gov.uk/swagger-ui/index.html#/agencies/getPrisonContactDetailList
https://api.prison.service.justice.gov.uk/swagger-ui/index.html#/agencies/getAgenciesByTypeUsingGET

Reply:
Ooooh good shout! I've updated our call to use the above endpoints, cheers! 👍